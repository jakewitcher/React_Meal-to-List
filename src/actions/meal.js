import uuid from 'uuid';

export const addMeal = (
    {
        name = '',
        itemList = [],
    } = {}
) => ({
    type: 'ADD_MEAL',
    meal: {
        id: uuid(),
        name,
        itemList,
    }
});

export const deleteMeal = (id = '') => ({
    type: 'DELETE_MEAL',
    id,
});

export const editMeal = (
    {
        name = '',
        id = '',
        itemList = [],
    } = {}
) => ({
    type: 'EDIT_MEAL',
    meal: {
        id,
        name,
        itemList,
    }
});