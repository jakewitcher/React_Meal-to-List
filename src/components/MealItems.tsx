import * as React from "react";
import MealItem from "./MealItem";
import { Item } from "../store/item/types";

interface IMealItemsProps {
  isDetails: boolean;
  itemList: Item[];
  mealName: string;
  deleteMealItem?(name: string): void;
  itemToEdit?(name: string): void;
}

const MealItems: React.FC<IMealItemsProps> = props => {
  const { mealName, itemList, deleteMealItem, itemToEdit, isDetails } = props;
  return (
    <div className="itemlist">
      <h1 className="itemlist__header">{mealName || "Meal"}</h1>
      {itemList.length > 0 && (
        <ul className="itemlist__body">
          {itemList.map(item => {
            return (
              <MealItem
                item={item}
                itemToEdit={itemToEdit}
                deleteMealItem={deleteMealItem}
                key={item.itemName}
                isDetails={isDetails}
              />
            );
          })}
        </ul>
      )}
    </div>
  );
};

export default MealItems;
