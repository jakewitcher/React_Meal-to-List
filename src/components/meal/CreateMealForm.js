import React from 'react';

const MealForm = (props) => {
    const {
        mealNameChange,
        itemNameChange,
        amountChange,
        unitChange,
        addItem,
        resetMeal,
        mealName,
        itemName,
        amount,
        unit,
        itemList,
        dispatch
    } = props.mealProps;
    const { addMeal } = props;

    return (
        <div className="form">
            <h2 className="form__header">Create a New Meal</h2>
            <div className="form__form">
                <div className="form__nameinputfield">
                    <p>Meal Name</p>
                    <input
                        className="form__nameinputfield--input"
                        type="text"
                        value={mealName}
                        onChange={mealNameChange}
                    />
                </div>
                <div className="form__iteminputfields">
                    <div>
                        <p>Item</p>
                        <input
                            className="form__iteminputfields--input"
                            type="text"
                            value={itemName}
                            onChange={itemNameChange}
                        />
                    </div>

                    <div>
                        <p>Amount</p>
                        <input
                            className="form__iteminputfields--input"
                            type="number"
                            value={amount}
                            onChange={amountChange}
                        />
                    </div>

                    <div>
                        <p>Unit</p>
                        <select
                            className="form__iteminputfields--input form__iteminputfields--dropdown"
                            name="unit"
                            value={unit}
                            onChange={unitChange}
                        >
                            <option value="pound(s)">pound(s)</option>
                            <option value="ounce(s)">ounce(s)</option>
                            <option value="gallon(s)">gallons</option>
                            <option value="quart(s)">quart(s)</option>
                            <option value="pint(s)">pint(s)</option>
                            <option value="item(s)">item(s)</option>
                            <option value="bag(s)">bag(s)</option>
                            <option value="box(es)">box(es)</option>
                            <option value="container(s)">container(s)</option>
                        </select>
                    </div>
                    <div className="form__button">
                        <button
                            className="form__button--add"
                            onClick={addItem}
                        >
                            Add Item
                        </button>
                    </div>
                </div>
                <div>
                    <button
                        className="form__button--save"
                        onClick={() => {
                            const name = mealName;
                            dispatch(addMeal({
                                name,
                                itemList
                            }));
                            resetMeal();
                        }}
                    >
                        Save Meal
                    </button>
                </div>
            </div>
        </div>
    )
}



export default MealForm;