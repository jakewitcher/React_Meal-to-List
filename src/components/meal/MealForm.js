import React from 'react';

const MealForm = (props) => {
    const {
        mealNameChange,
        itemNameChange,
        amountChange,
        unitChange,
        addItem,
        resetMeal,
        mealName,
        itemName,
        amount,
        unit,
        itemList,
        updateMeal,
        mealId,
        title,
    } = props;

    const pluralUnitValue = (amount, unitType) => {
        if (amount < 2) {
            return `${unitType}`
        }
        return `${unitType}s`
    };

    return (
        <div className="form">
            <h2 className="form__header">{title}</h2>
            <div className="form__form">
                <div className="form__nameinputfield">
                    <p>Meal Name</p>
                    <input
                        className="form__nameinputfield--input"
                        type="text"
                        value={mealName}
                        onChange={mealNameChange}
                    />
                </div>
                <div className="form__iteminputfields">
                    <div>
                        <p>Item</p>
                        <input
                            className="form__iteminputfields--input"
                            type="text"
                            value={itemName}
                            onChange={itemNameChange}
                        />
                    </div>

                    <div>
                        <p>Amount</p>
                        <input
                            className="form__iteminputfields--input"
                            type="number"
                            value={amount}
                            onChange={amountChange}
                        />
                    </div>

                    <div>
                        <p>Unit</p>
                        <select
                            className="form__iteminputfields--input form__iteminputfields--dropdown"
                            name="unit"
                            value={unit}
                            onChange={unitChange}
                        >
                            <option value={pluralUnitValue(amount, 'pound')}>{pluralUnitValue(amount, 'pound')}</option>
                            <option value={pluralUnitValue(amount, 'ounce')}>{pluralUnitValue(amount, 'ounce')}</option>
                            <option value={pluralUnitValue(amount, 'gallon')}>{pluralUnitValue(amount, 'gallon')}</option>
                            <option value={pluralUnitValue(amount, 'quart')}>{pluralUnitValue(amount, 'quart')}</option>
                            <option value={pluralUnitValue(amount, 'pint')}>{pluralUnitValue(amount, 'pint')}</option>
                            <option value={pluralUnitValue(amount, 'item')}>{pluralUnitValue(amount, 'item')}</option>
                            <option value={pluralUnitValue(amount, 'bag')}>{pluralUnitValue(amount, 'bag')}</option>
                            <option value={pluralUnitValue(amount, 'container')}>{pluralUnitValue(amount, 'container')}</option>
                        </select>
                    </div>
                    <div className="form__button">
                        <button
                            className="form__button--add"
                            onClick={addItem}
                        >
                            Add Item
                        </button>
                    </div>
                </div>
                <div>
                    <button
                        className="form__button--save"
                        onClick={() => {
                            const name = mealName;
                            const id = mealId;
                            updateMeal({
                                name,
                                itemList,
                                id,
                            });
                            resetMeal();
                        }}
                    >
                        Save Meal
                    </button>
                </div>
            </div>
        </div>
    )
}

export default MealForm;