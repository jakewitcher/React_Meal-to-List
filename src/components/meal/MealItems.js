import React from 'react';
import MealItem from './MealItem'

const MealItems = (props) => {
    const { mealName, itemList, deleteMealItem, itemToEdit } = props;
    return (
        <div className="itemlist">
            <h1 className="itemlist__header">{mealName || 'Meal'}</h1>
            {itemList.length > 0 && <ul className="itemlist__body">
                {itemList.map(item => {
                    return (
                        <MealItem
                            item={item}
                            itemToEdit={itemToEdit}
                            deleteMealItem={deleteMealItem}
                            key={item.itemName}
                        />
                    )
                })}
            </ul>}
        </div>
    );
}

export default MealItems;