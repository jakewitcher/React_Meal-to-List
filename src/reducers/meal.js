import { defaultMealsList } from './seed';

// const mealDefaultState = {
//     mealList: [],
// };

export default (state = {mealList: defaultMealsList}, action) => {
    switch (action.type) {
        case 'ADD_MEAL':
            return {
                ...state,
                mealList: [...state.mealList, action.meal]
            };
        default:
            return state;
    }
}