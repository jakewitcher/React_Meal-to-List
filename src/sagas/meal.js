import { put, takeEvery, select } from 'redux-saga/effects';
import database from '../firebase/firebase';

const getUserId = (state) => state.auth.uid;

const formatItemList = (list) => {
    return list.reduce((list, item) => {
        list[item.itemName] = {
            amount: item.amount,
            unit: item.unit
        };
        return list;
    }, {});
};

function* addMeal({ meal = {} } = {}) {
    const { itemList, name } = meal;
    const items = formatItemList(itemList);

    let id;
    const uid = yield select(getUserId);
    yield database.ref(`users/${uid}/meals`).push({ name, items }).then((ref) => {
        id = ref.key;
    });

    yield put({
        type: 'ADD_MEAL',
        meal: {
            name,
            itemList,
            id,
        }
    });
}

function* editMeal({ meal = {} }) {
    const { itemList, name, id } = meal;
    const items = formatItemList(itemList);
    const uid = yield select(getUserId);

    yield database.ref(`users/${uid}/meals/${id}`).update({ name, items });

    yield put({
        type: 'EDIT_MEAL',
        meal: {
            name,
            itemList,
            id,
        }
    });
};

function* deleteMeal({ id = '' }) {
    const uid = yield select(getUserId);

    yield database.ref(`users/${uid}/meals/${id}`).remove();

    yield put({
        type: 'DELETE_MEAL',
        id,
    });
}

function* setMeal() {
    const meals = [];
    const uid = yield select(getUserId);

    yield database.ref(`users/${uid}/meals`).once('value').then((snapshot) => {
        snapshot.forEach((childSnapshot) => {
            const itemList = [];
            childSnapshot.child('items').forEach((item) => {
                itemList.push({
                    itemName: item.key,
                    amount: item.val().amount,
                    unit: item.val().unit,
                });
            });

            meals.push({
                id: childSnapshot.key,
                name: childSnapshot.val().name,
                itemList,
            });
        });
    });

    yield put({ type: 'SET_MEAL', meals });
};

export function* watchAddMeal() {
    yield takeEvery('ON_ADD_MEAL', addMeal);
};

export function* watchEditMeal() {
    yield takeEvery('ON_EDIT_MEAL', editMeal);
};

export function* watchDeleteMeal() {
    yield takeEvery('ON_DELETE_MEAL', deleteMeal);
};

export function* watchSetMeal() {
    yield takeEvery('ON_SET_MEAL', setMeal);
}