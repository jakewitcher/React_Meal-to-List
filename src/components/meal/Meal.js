import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { onDeleteMeal } from '../../actions/meal';

export const Meal = (props) => {
    const { name, id, itemList } = props;
    return (
        <div className="list-type">
            <Link className="list-type__link" to={`/meals/meal/${id}`}>
                <p className="list-type__name">{name}</p>
            </Link>
            <p className="list-type__details">number of items: <span className="list-type__details list-type__details--span"> {itemList.length}</span></p>
            <div className="list-type__update">
                <Link className="list-type__link" to={`/meals/edit/${id}`}>
                    <i className="list-type__edit fas fa-edit" />
                </Link>
                <i
                    className="list-type__delete fas fa-trash"
                    onClick={() => props.onDeleteMeal(id)}
                />
            </div>
        </div>
    );
}

const mapDispatchToProps = (dispatch) => {
    return {
        onDeleteMeal: (id) => dispatch(onDeleteMeal(id)),
    }
}

export default connect(undefined, mapDispatchToProps)(Meal);