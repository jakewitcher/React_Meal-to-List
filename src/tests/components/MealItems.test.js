import React from 'react';
import { shallow } from 'enzyme';
import MealItems from '../../components/meal/MealItems';
import { mealsList } from '../fixtures/meal';

test('should render MealItems correctly', () => {
    const mealName = mealsList[0].name;
    const itemList = mealsList[0].itemList;
    const wrapper = shallow(<MealItems mealName={mealName} itemList={itemList} />);
    expect(wrapper).toMatchSnapshot();
});

