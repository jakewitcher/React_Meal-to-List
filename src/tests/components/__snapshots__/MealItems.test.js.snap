// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MealItems correctly 1`] = `
<div
  className="itemlist"
>
  <h1
    className="itemlist__header"
  >
    breakfast
  </h1>
  <ul
    className="itemlist__body"
  >
    <MealItem
      item={
        Object {
          "amount": 2,
          "id": 1,
          "itemName": "bacon",
          "unit": "pounds",
        }
      }
      key="bacon"
    />
    <MealItem
      item={
        Object {
          "amount": 12,
          "id": 2,
          "itemName": "eggs",
          "unit": "items",
        }
      }
      key="eggs"
    />
    <MealItem
      item={
        Object {
          "amount": 1,
          "id": 3,
          "itemName": "toast",
          "unit": "bag",
        }
      }
      key="toast"
    />
    <MealItem
      item={
        Object {
          "amount": 1,
          "id": 4,
          "itemName": "orange juice",
          "unit": "quart",
        }
      }
      key="orange juice"
    />
  </ul>
</div>
`;
